### PR TITLE
Support Azure Application Gateway

### LINKED ISSUE #191: Support Application Gateway
UI design for supporting Azure Application Gateway:

## Screenshots

* image
* image

## UI changes summary

* In the renamed **Load balancer** tab:
   * The tab name is changed from **IBM HTTP Server Load Balancer** to **Load balancer**
   * The name of options group is updated to **Select a load balancer**. Its tooltip is also updated accordingly with the new load balancer option **Azure Application Gateway**
   * Use options group to allow user to select a load balancer (Azure Application Gateway / IBM HTTP Server) or not
   * Provide Azure Application Gateway settings for user inputs if user selects to deploy Azure Application Gateway 
* In the **Networking** tab
  * Update the description in the information box if user selects to deploy Azure Application Gateway
  * Added a new subnet for Application Gateway if user selects to deploy Azure Application Gateway

## Preview UI changes 

Reviewer can also preview the UI changes using the Create UI Definition Sandbox:

* Open [Create UI Definition Sandbox](https://portal.azure.com/?feature.customPortal=false#blade/Microsoft_Azure_CreateUIDef/SandboxBlade) in a new tab of browser.
* Open UI source code [createUiDefinition.json](https://raw.githubusercontent.com/majguo/azure.websphere-traditional.cluster/support-app-gw/src/main/arm/createUiDefinition.json) in a new tab of the browser.
* Replace the content of work area in **Create UI Definition Sandbox** with the content of UI source code `createUiDefinition.json`.
* Click Preview at the left-bottom of **Create UI Definition Sandbox**.

### PR DESCRIPTION
**Pull request by majguo**
## Description

The PR is to resolve the following issues:
* Resolve #191 

## Change summary

* Add a new load balancer option **Azure Application Gateway** in UI
* Define uami, key-vault with TLS certificate and Azure Application Gateway in Bicep modules
* Add selecting load balancer option (IBM HTTP Server, Azure Application Gateway and none) into `integration-test` pipeline

## Testing

Verified the following 3 test cases with `integration-test` pipeline:
|Load balancer |BYOV |Database connection |Pipeline run|
|----------------|-------|------------------------|-------------|
|IBM HTTP Server|No|Microsoft SQL Server|[integration-test-89](https://github.com/majguo/azure.websphere-traditional.cluster/actions/runs/3736582366)|
|Azure Application Gateway|No|Microsoft SQL Server|[integration-test-92](https://github.com/majguo/azure.websphere-traditional.cluster/actions/runs/3736936664)|
|None|No|Microsoft SQL Server|[integration-test-91](https://github.com/majguo/azure.websphere-traditional.cluster/actions/runs/3736583226)|

If you want to enable Azure Application Gateway in `integration-test` pipeline, pls look at #194.

Verified the following 6 test cases using the [preview link](https://portal.azure.com/?feature.customportal=false#create/microsoft_javaeeonazure_test.20191212-arm-rhel-was-nd-v9-cluster-previewwasndt91-n-cluster):
|Load balancer |BYOV |Database connection |
|----------------|-------|------------------------|
|IBM HTTP Server|No|Microsoft SQL Server|
|Azure Application Gateway|No|Microsoft SQL Server|
|None|No|Microsoft SQL Server|
|IBM HTTP Server|Yes|Microsoft SQL Server|
|Azure Application Gateway|Yes|Microsoft SQL Server|
|None|Yes|Microsoft SQL Server|

Feel free to do a live testing with the [preview link](https://portal.azure.com/?feature.customportal=false#create/microsoft_javaeeonazure_test.20191212-arm-rhel-was-nd-v9-cluster-previewwasndt91-n-cluster).

## UI changes review

Besides visiting [preview link](https://portal.azure.com/?feature.customportal=false#create/microsoft_javaeeonazure_test.20191212-arm-rhel-was-nd-v9-cluster-previewwasndt91-n-cluster) for reviewing the new database type `Oracle database` in blade `Database`, reviewer can preview the UI changes using the **Create UI Definition Sandbox**:

- Open [Create UI Definition Sandbox](https://portal.azure.com/?feature.customPortal=false#blade/Microsoft_Azure_CreateUIDef/SandboxBlade) in a new tab of browser.
- Open UI source code [createUiDefinition.json](https://raw.githubusercontent.com/majguo/azure.websphere-traditional.cluster/support-app-gw/src/main/arm/createUiDefinition.json) in a new tab of the browser.
- Replace the content of work area in **Create UI Definition Sandbox** with the content of **UI source code `createUiDefinition.json`**.
- Click **Preview** at the left-bottom of **Create UI Definition Sandbox**.

### UI changes summary

* In the renamed **Load balancer** tab:
   * The tab name is changed from **IBM HTTP Server Load Balancer** to **Load balancer**
   * The name of options group is updated to **Select a load balancer**. Its tooltip is also updated accordingly with the new load balancer option **Azure Application Gateway**
   * Use options group to allow user to select a load balancer (Azure Application Gateway / IBM HTTP Server) or not
   * Provide Azure Application Gateway settings for user inputs if user selects to deploy Azure Application Gateway 
* In the **Networking** tab
  * Update the description in the information box if user selects to deploy Azure Application Gateway
  * Added a new subnet for Application Gateway if user selects to deploy Azure Application Gateway

### Screenshots

* image
* image

Signed-off-by: Jianguo Ma <jiangma@microsoft.com>

### CODE REVIEW

File: .github/workflows/integration-test.yaml [hunk 1]
```diff
@@ -225,10 +241,40 @@ jobs:
...
+          fi
+          appGatewayHttpURL=$(echo $outputs | jq -r '.appGatewayHttpURL.value')
+          appGatewayHttpsURL=$(echo $outputs | jq -r '.appGatewayHttpsURL.value')
+          if [[ $selectLoadBalancer == "appgw" ]]; then
+            curl $appGatewayHttpURL
+            if [[ $? -ne 0 ]]; then
+              echo "Failed to access ${appGatewayHttpURL}."
+              exit 1
+            fi
+            curl $appGatewayHttpsURL -k
```
Comment (edburns) [inline .github/workflows/integration-test.yaml:264]: It is probably safer to add a timeout and retry. Because there are multiple curl commands, define the values in the environment section so you can change them easily if desired.

File: src/main/bicep/modules/_azure-resources/_appGateway.bicep [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+            name: 'LocationHeader${i}'
+            ruleSequence: 50
+            conditions: [
+              {
+                variable: 'http_resp_Location'
+                pattern: format('(https?):\\/\\/{0}:{1}(.*)$', reference(resourceId('Microsoft.Network/networkInterfaces', '${workerNodePrefix}${i}-if'), '2021-08-01').ipConfigurations[0].properties.privateIPAddress, const_backendPort)
+                ignoreCase: true
+                negate: false
+              }
+            ]
```
Comment (edburns) [inline src/main/bicep/modules/_azure-resources/_appGateway.bicep:231]: Great job catching this corner case!

**Review by edburns — CHANGES_REQUESTED** [review]
Other than the retries in the `integration-test.yaml`, this looks very good.  Please add the retry and timeout and I'll approve it.

File: .github/workflows/integration-test.yaml (hunk 1 above)
Comment (majguo) [reply .github/workflows/integration-test.yaml]: Good catch, fixed in [41ccb83](https://github.com/WASdev/azure.websphere-traditional.cluster/pull/193/commits/41ccb8395a12b0dbb856bf052857fc8721a5eb77).

**Comment by majguo** [comment]
> Other than the retries in the `integration-test.yaml`, this looks very good. Please add the retry and timeout and I'll approve it.

Fixed in [41ccb83](https://github.com/WASdev/azure.websphere-traditional.cluster/pull/193/commits/41ccb8395a12b0dbb856bf052857fc8721a5eb77).

**Review by edburns — APPROVED** [review]
